torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -138,53 +132,56 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
     future->markCompleted();
   }
 
-  enqueue(SendWork(dstRank, std::move(message)));
+  enqueueSend(RpcWork(dstRank, std::move(message)));
   return future;
 }
 
-void ProcessGroupAgent::enqueue(SendWork work) {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  sendQueue_.emplace_back(std::move(work));
-  lock.unlock();
-
-  workProduceCV_.notify_one();
+void ProcessGroupAgent::enqueueSend(RpcWork work) {
+  threadPool_.run(std::bind(
+    [&](const RpcWork& work) {
+      std::stringstream ss;
+      serialize(work.message_, ss);
+      std::string str = ss.str();
+
+      std::vector<torch::Tensor> preamble = {
+        torch::tensor(
+          {
+            (int64_t)pg_->getRank(),
+            (int64_t)str.length(),
+          }, {torch::kLong})
+      };
+
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
+      std::unique_lock<std::mutex> lock(sendMutexes_.get()[work.rank_]);
+      pg_->send(preamble, work.rank_, work.rank_ /* channelTag */)->wait();
+      std::vector<torch::Tensor> payload =
+          {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
+      pg_->send(payload, work.rank_, work.rank_ /* channelTag */)->wait();
+      lock.unlock();
+    },
+    std::move(work)
+  ));
 }
 
-// making sure tensors are not deleted before send finishes
-void ProcessGroupAgent::sendLoop() {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-
-  while (!stop_) {
-    if (sendQueue_.empty()) {
-      workProduceCV_.wait(lock);
-      continue;
-    }
-
-    auto work = std::move(sendQueue_.front());
-    sendQueue_.pop_front();
-    lock.unlock();
-
-    workConsumeCV_.notify_one();
-
-
-    std::stringstream ss;
-    serialize(work.message_, ss);
-    std::string str = ss.str();
-
-    std::vector<torch::Tensor> preamble = {
-      torch::tensor(
+void ProcessGroupAgent::enqueueRecv(RpcWork work) {
+  threadPool_.run(std::bind(
+    [&](RpcWork& work) {
+      if (work.message_.isRequest()) {
+        cb_(names_[work.rank_], std::move(work.message_), *this);
+      } else if (work.message_.isResponse()) {
+        auto id = work.message_.id();
         {
-          (int64_t)pg_->getRank(),
-          (int64_t)str.length(),
-        }, {torch::kLong})
-    };
-    pg_->send(preamble, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-    std::vector<torch::Tensor> payload =
-        {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
-    pg_->send(payload, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-
-    lock.lock();
-  }
+          std::lock_guard<std::mutex> lock{futureMutex_};
+          futures_[id]->markCompleted(std::move(work.message_));
+          futures_.erase(id);
+        }
+      } else {
+        AT_ERROR("unrecognized message type ", work.message_.type());
```

Comment:
This causes an exception to be thrown, but there is no-one to handle it. If it is really an unrecognized message, shouldn't this be a `TORCH_INTERNAL_ASSERT` and have it crash the process?

Reply:
How about adding error handling to `FutureMessage` (or just use `ivalue::Future`'s error handle after we make Message an iValue)? So the error will be passed to the user when the sync rpc returns or `FutureMessage::wait()` returns.

Reply:
Sounds good. As long as we always have a message to associate this error with, of course. Ideally, this function should be marked noexcept for that reason, to force all errors to propagate back to the user.